examples/ethernet/ethernet/main/ethernet_main.c

```diff
@@ -32,92 +32,45 @@
 #include "tcpip_adapter.h"
 #include "nvs_flash.h"
 #include "driver/gpio.h"
-#include "tlk110_phy.h"
 
-static const char *TAG = "eth_demo";
+extern void phy_device_check_phy_init(void);
```

Comment:
Feel free to disagree with me here, but I think rather than declare all these functions as extern, it may be better to make them static in the individual source files and then declare inside each PHY-specific file

```
const eth_config_t tlk110_default_ethernet_phy_config = {
    .phy_check_init = phy_tlk110_check_phy_init,
    .phy_check_link = phy_tlk110_check_phy_link_status,
    .phy_get_speed_mode = phy_tlk110_get_speed_mode,
    .phy_get_duplex_mode = phy_tlk110_get_duplex_mode,
    .flow_ctrl_enable = true;
    .phy_get_partner_pause_enable = phy_tlk110_get_partner_pause_enable,
etc}
```

And then in this code:

```
extern const eth_config_t tlk110_default_ethernet_phy_config;
```

Which then be either passed to the esp_eth_init() directly (if the signature is changed to const), or copied to another structure which is then passed in directly.

This also has the advantage that you don't need to selectively compile the source files, we can compile both but only one will be linked into the final binary depending on which extern-ed config structure is used.

(Sorry if this seems different to what I was suggesting before, I realised I gave two different recommendations as I read my way through the code.)